cmd/restic/cmd_cleanup_packs.go

```diff
@@ -0,0 +1,330 @@
+		notFinalIdx := (repo.Index()).(*repository.MasterIndex).NotFinalIndexes()
+		if len(notFinalIdx) != 1 {
+			return errors.Fatal("should only have one unfinalized index!!")
+		}
+		repackIndex := notFinalIdx[0]
+
+		err = ChangePacksInIndex(opts, gopts, repo, repackIndex)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(removePacks) != 0 {
+		Verbosef("deleting packs (unused and repacked)\n")
+		fileHandles := make(chan restic.Handle)
+		go func() {
+			for packID := range removePacks {
+				fileHandles <- restic.Handle{Type: restic.DataFile, Name: packID.String()}
+			}
+			close(fileHandles)
+		}()
+		DeleteFiles(gopts, opts.DryRun, repo, len(removePacks), fileHandles)
+
+	}
+
+	Verbosef("done\n")
+	return nil
+}
+
+const numDeleteWorkers = 8
+
+func DeleteFiles(gopts GlobalOptions, dryrun bool, repo restic.Repository, totalCount int, fileHandles <-chan restic.Handle) {
+	var wg sync.WaitGroup
+	bar := newProgressMax(!gopts.Quiet, uint64(totalCount), "files deleted")
+	bar.Start()
+	wg.Add(numDeleteWorkers)
+	for i := 0; i < numDeleteWorkers; i++ {
+		go func() {
+			for h := range fileHandles {
+				if !dryrun {
+					err := repo.Backend().Remove(gopts.ctx, h)
+					if err != nil {
+						Warnf("unable to remove file %v from the repository\n", h.Name)
+					}
+					if !gopts.JSON && gopts.verbosity >= 2 {
+						Verbosef("%v was removed.\n", h.Name)
+					}
+				} else {
+					if !gopts.JSON && gopts.verbosity >= 2 {
+						Verbosef("would have removed %v.\n", h.Name)
+					}
+				}
+				bar.Report(restic.Stat{Blobs: 1})
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+	bar.Done()
+}
+
+func repack(ctx context.Context, repo restic.Repository, blobs []restic.Blob, p *restic.Progress) (err error) {
+
+	var buf []byte
+	for _, entry := range blobs {
+		// if blob is not in index, don't write it
+		if !repo.Index().Has(entry.ID, entry.Type) {
+			debug.Log("not writing blob %v (not in index)", entry.ID)
+			continue
+		}
+
+		buf = buf[:]
+		if uint(len(buf)) < entry.Length {
+			buf = make([]byte, entry.Length)
+		}
+		buf = buf[:entry.Length]
+
+		// TODO: Use encrypted blob and just save it instead decrypting and encrypting again
+		n, err := repo.LoadBlob(ctx, entry.Type, entry.ID, buf)
+
+		if err != nil {
+			return errors.Wrap(err, "LoadBlob")
+		}
+		if n+crypto.Extension != int(entry.Length) {
+			return errors.Errorf("error reading blob %v; length invalid. Want %d, got %d", entry.ID, entry.Length, n)
+		}
+
+		buf = buf[:n]
+		_, err = repo.SaveBlob(ctx, entry.Type, buf, entry.ID)
+		if err != nil {
+			return err
+		}
+		debug.Log("  saved blob %v", entry.ID)
+	}
+
+	if p != nil {
+		p.Report(restic.Stat{Blobs: 1})
+	}
+	return nil
+}
+
+func ChangePacksInIndex(opts CleanupPacksOptions, gopts GlobalOptions, repo restic.Repository, repackIndex *repository.Index) error {
+	return ModifyIndex(opts.DryRun, gopts, repo, func(pb restic.PackedBlob) (changed bool, pbnew restic.PackedBlob) {
+		pbs, found := repackIndex.Lookup(pb.ID, pb.Type)
```

Comment:
Thanks for this hint! I agree that (partly) rebuilding the index is the better solution. A "modification" also is just a "save new version" and "delete old one". Hence I'll change the code that it will be a true rebuild.